File: Backend/model/src/main/java/io/sparkled/model/animation/fill/reference/FillTypes.java
package io.sparkled.model.animation.fill.reference;

import io.sparkled.model.animation.fill.FillType;
import io.sparkled.model.animation.param.Param;
import io.sparkled.model.animation.param.ParamType;
import io.sparkled.model.animation.fill.FillTypeCode;
import io.sparkled.model.animation.param.ParamName;

import java.util.Arrays;
import java.util.List;

public class FillTypes {
    private static final List<FillType> TYPES = Arrays.asList(
            fill(FillTypeCode.RAINBOW, "Rainbow",
                    param(ParamName.CYCLE_COUNT, ParamType.DECIMAL).setValue(1),
                    param(ParamName.CYCLES_PER_SECOND, ParamType.DECIMAL).setValue(1)
            ),
            fill(FillTypeCode.SOLID, "Solid",
                    param(ParamName.COLOR, ParamType.COLOR)
            )
    );

    public static List<FillType> get() {
        return TYPES;
    }

    private static FillType fill(FillTypeCode fillType, String name, Param... params) {
        return new FillType().setCode(fillType).setName(name).setParams(Arrays.asList(params));
    }

    private static Param param(ParamName paramName, ParamType type) {
        return new Param().setName(paramName.getName()).setType(type);
    }
}

File: model/src/main/kotlin/io/sparkled/model/animation/fill/reference/FillTypes.kt
package io.sparkled.model.animation.fill.reference

import io.sparkled.model.animation.fill.FillType
import io.sparkled.model.animation.fill.FillTypeCode
import io.sparkled.model.animation.param.Param
import io.sparkled.model.animation.param.ParamCode
import io.sparkled.model.animation.param.ParamType
import io.sparkled.model.util.ParamUtils.param
import java.util.Arrays

object FillTypes {
    private val TYPES = Arrays.asList(
        fill(
            FillTypeCode.GRADIENT,
            param(ParamCode.COLORS, ParamType.COLORS, "#ff0000", "#0000ff"),
            param(ParamCode.COLOR_REPETITIONS, ParamType.INTEGER, 1),
            param(ParamCode.BLEND_HARDNESS, ParamType.DECIMAL, 0),
            param(ParamCode.CYCLES_PER_SECOND, ParamType.DECIMAL, 0)
        ),
        fill(
            FillTypeCode.RAINBOW,
            param(ParamCode.BRIGHTNESS, ParamType.DECIMAL, 100),
            param(ParamCode.CYCLE_COUNT, ParamType.DECIMAL, 1),
            param(ParamCode.CYCLES_PER_SECOND, ParamType.DECIMAL, .5f)
        ),
        fill(
            FillTypeCode.SOLID,
            param(ParamCode.COLOR, ParamType.COLOR, "#ff0000")
        )
    )

    fun get(): List<FillType> {
        return TYPES
    }

    private fun fill(code: FillTypeCode, vararg params: Param): FillType {
        return FillType(code, code.displayName, listOf(*params))
    }
}
